src/EditorFeatures/CSharp/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProvider.cs

```diff
@@ -0,0 +1,394 @@
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.CSharp.MisplacedUsingDirectives
+{
+    /// <summary>
+    /// Implements a code fix for all misplaced using statements.
+    /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.MoveMisplacedUsings)]
+    [Shared]
+    internal sealed partial class MisplacedUsingDirectivesCodeFixProvider : CodeFixProvider
+    {
+        private static readonly SyntaxAnnotation s_usingPlacementCodeFixAnnotation = new SyntaxAnnotation(nameof(s_usingPlacementCodeFixAnnotation));
+
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.MoveMisplacedUsingsDiagnosticId);
+
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return FixAll.Instance;
+        }
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var syntaxRoot = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var compilationUnit = (CompilationUnitSyntax)syntaxRoot;
+            var options = await context.Document.GetOptionsAsync(context.CancellationToken).ConfigureAwait(false);
+
+            // Do not offer a code fix when there are multiple namespaces in the source file.
+            if (CountNamespaces(compilationUnit.Members) > 1)
+            {
+                return;
+            }
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MoveMisplacedUsingsCodeAction(cancellationToken => GetTransformedDocumentAsync(context.Document, compilationUnit, options, cancellationToken)),
+                    diagnostic);
+            }
+        }
+
+        private static async Task<Document> GetTransformedDocumentAsync(Document document, CompilationUnitSyntax compilationUnit, OptionSet options, CancellationToken cancellationToken)
+        {
+            var usingDirectivesPlacement = DeterminePlacement(compilationUnit, options);
+
+            // There should only be a diagnostic when there are usings directives that need moving.
+            Debug.Assert(!usingDirectivesPlacement.preferPreservation);
+
+            var compilationUnitWithExpandedUsings = await ExpandUsingDirectivesAsync(document, compilationUnit, cancellationToken).ConfigureAwait(false);
+
+            var (compilationUnitWithoutHeader, fileHeader) = RemoveFileHeader(compilationUnitWithExpandedUsings);
+
+            var newCompilationUnit = usingDirectivesPlacement.placement switch
+            {
+                AddImportPlacement.InsideNamespace => MoveUsingsInsideNamespace(compilationUnitWithoutHeader),
+                AddImportPlacement.OutsideNamespace => MoveUsingsOutsideNamespace(compilationUnitWithoutHeader),
+                _ => throw ExceptionUtilities.Unreachable
+            };
```

Comment:
if you only have two values, why not use a `?:`

Reply:
moving usings inside/outside a namespace can cause code to change meaning (or produce errors).  Are we doing anything to prevent that?  i.e. should we expand/reduce all types in the affected scope?  or are we just punting and saying that's an acceptable fallout of the change?  does hte change warn at all that moving scopes may affect the code?

Should we do a prepass on the referenced types to see if tehre might be any ambiguity with the usings vs the namespace they're currently in, and only warn then?

Reply:
The ambiguity check you mentioned would be quite expensive though. I think adding a warning would suffice.

Reply:
@genlu's likely right.  The file could be effectively unbounded in size.  I think a simple warning annotation woud be fine.

Reply:
Added a warning annotation to the moved using directives. Unfortunately the improvement to preview warnings (https://github.com/dotnet/roslyn/pull/34485) won't be in until preview 3.